build: ensure only errors are displayed in issue via log file

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds support for an `ESLINT_ERROR_LOG` variable to store errors when running `eslint-files` make rule.

Using the JSON formatter as suggested in the issue and discussed doesn't help much since we process each file individually and would then still need to parse out the results for them. Instead, I ended up settling on re-running the ESLint command with `--quiet` to ensure only errors get reported and appending the results to a log file whenever linting of a file fails.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/45.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add error log variable to store errors in file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
> Using the JSON formatter as suggested in the issue and discussed doesn't help much since we process each file individually and would then still need to parse out the results for them

Why wasn't the Unix formatter suitable?

Planeshifter commented:
@kgryte 

If we used the Unix formatter, we would get output like

> /var/lib/jenkins/workspace/eslint Release/eslint/fullOfProblems.js:1:10: 'addOne' is defined but never used. [Error/no-unused-vars]

We could implement this and then use `grep` to filter lines containing `[Error`. However, I find this output less appealing than the default formatting. Second, we would need to maintain the previous `tee` approach in order to print to both stdout and a file then. Finally, we would need to either use a deprecated ESLint formatter (no longer available in v9) or add that external dependency. In comparison, I find the `make` rule variable for storing errors simple and clear.

kgryte commented on tools/make/lib/lint/javascript/eslint.mk:
```diff
@@ -44,6 +44,9 @@ ESLINT_CONF_BENCHMARKS ?= $(CONFIG_DIR)/eslint/.eslintrc.benchmarks.js
+ESLINT_ERROR_LOG ?=
+
 # Define the command-line options to use when invoking the ESLint executable:
 ESLINT_FLAGS ?= \
```
I would suggest converting this to an internal variable `eslint_flags` and then creating a separate `ESLINT_FLAGS ?=` which is user defined. Then you can concat the two variables and use `eslint_flags` below.

We do this elsewhere in various make recipes.

kgryte commented on tools/make/lib/lint/javascript/eslint.mk:
```diff
@@ -251,6 +254,18 @@ ifeq ($(FAIL_FAST), true)
+	for file in $(FILES); do \
+		echo ''; \
+		echo "Linting file: $$file"; \
+		if ! $(ESLINT) $(ESLINT_FLAGS) --config $(ESLINT_CONF) $$file; then \
```
I suggest moving this retry logic to the Yaml file.

kgryte commented on .github/workflows/lint_random_files.yml:
```diff
@@ -335,31 +335,31 @@ jobs:
 
           if [[ -n "${files}" ]]; then
-            make lint-javascript-files FIX="${FIX}" FAST_FAIL=0 FILES="${files}" 2>&1 | tee -a "$ERR_FILE"
+            make lint-javascript-files FIX="${FIX}" FAST_FAIL=0 FILES="${files}" ESLINT_ERROR_LOG="$ERR_FILE"
```
To avoid redundancy, I would suggest creating a small helper function which performs the retry logic and then redirects to `ERR_FILE` on lint failure.

kgryte commented on .github/workflows/lint_random_files.yml:
```diff
@@ -335,31 +335,31 @@ jobs:
           done
```
I suggest moving this logic to a separate Bash script. This will make refactoring and linting easier.

Planeshifter committed:
chore: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@Planeshifter Would it be possible to move this PR along? Would help to minimize some of the PR noise we are experiencing for the lint issues.

Planeshifter committed:
fix: correct Makefile syntax and variable usage in eslint rules

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
build: move error log linting step to script file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on .github/workflows/scripts/lint_javascript_files_error_log:
```diff
@@ -0,0 +1,174 @@
+# MAIN #
+
+# Initialize error file:
+: > "${err_file}"
```
This is valid and a trick, but not typical. One would conventionally use `touch` here.

kgryte commented on .github/workflows/scripts/lint_javascript_files_error_log:
```diff
@@ -0,0 +1,174 @@
+files=$(echo "${files_to_lint}" | tr ' ' '\n' | grep '\.js$' | grep -v -e '/examples' -e '/test' -e '/benchmark' -e '^dist/' | tr '\n' ' ' | sed 's/ $//')
+if [ -n "${files}" ]; then
+	echo "Linting JavaScript source files..."
+	run_make_lint "${files}" "" ""
```
This parameterization of `run_make_lint` is a bit odd and not idiomatic. This isn't the end of the world, but the implementation could be cleaner.

kgryte commented on .github/workflows/scripts/lint_javascript_files_error_log:
```diff
@@ -0,0 +1,174 @@
+files=$(echo "${files_to_lint}" | tr ' ' '\n' | grep '/examples/.*\.js$' | tr '\n' ' ' | sed 's/ $//')
+if [ -n "${files}" ]; then
+	echo "Linting JavaScript example files..."
+	run_make_lint "${files}" "" "ESLINT_CONF=\"${eslint_examples_conf}\""
```
E.g., escaping environment variable strings, etc, is odd.
